File: src/compilers/java.ts
import { spawn } from 'child_process';

const javaCompiler = (input: string, file: string) => {
    return new Promise((resolve, reject) => {
        const javaProcess = spawn('java', [file]);
        let time = Date.now();
        let outputData = '';
        setTimeout(() => {
            javaProcess.kill();
            resolve({
                success: true,
                error: true,
                message: 'Execution Time Out',
                data: outputData.toString(), time: 3000
            });
        }, 3000);

        javaProcess.stdin.write(input);
        javaProcess.stdin.end();

        javaProcess.stdout.on('data', (data) => {
            outputData += data;
        });

        let isError: boolean = false;
        javaProcess.on('close', () => {
            resolve({
                success: true,
                error: isError,
                message: 'Compilation Error',
                data: outputData.toString(),
                time: Date.now() - time
            });
        });

        javaProcess.stderr.on('data', (data) => {
            isError = true;
            outputData += data;
        });
    });
}

export default javaCompiler;

File: goserver/compilers/java.go
package compilers

import (
	"bytes" 
	"os"
	"os/exec"
	"time"

	"github.com/deepaksuthar40128/plumber/utils"
)

func JAVACompiler(f *os.File, inputf *os.File)utils.OutgoingDataType {
	defer func() {
		f.Close()
		inputf.Close()
		utils.RemoveFile(f)
		utils.RemoveFile(inputf)
	}()
	cmd := exec.Command("java", f.Name())
	var stdout bytes.Buffer
	cmd.Stderr = &stdout
	cmd.Stdout = &stdout
	inputf.Close()
	inputf, err := os.Open(inputf.Name())
	if err != nil {
		panic("Error during reopening input file")
		
	}
	cmd.Stdin = inputf
	output := make(chan utils.OutgoingDataType)
	go func() {
		nchannel := make(chan utils.OutgoingDataType)
		go func() {
			startTime := time.Now()
			if err := cmd.Run(); err != nil { 
				nchannel <- utils.OutgoingDataType{
					Success:    true,
					Error:      true,
					Message:    "Execution Terminated Due to Some Error",
					Data:       utils.BufferOverflowCheck(&stdout),
					Time:       0,
					StatusCode: 200,
				}
			}
			nchannel <- utils.OutgoingDataType{
				Success:    true,
				Error:      false,
				Message:    "Run successfully",
				Data:       utils.BufferOverflowCheck(&stdout),
				Time:       int(time.Since(startTime) / 1e6),
				StatusCode: 200,
			}
		}()
		select {
		case res := <-nchannel:
			output <- res
		case <-time.After(2 * time.Second):
			output <- utils.OutgoingDataType{
				Success:    true,
				Error:      true,
				Message:    "Time Limit Exceeded",
				Data:       utils.BufferOverflowCheck(&stdout),
				Time:       2000,
				StatusCode: 200,
			}
			cmd.Process.Kill()
		}
	}()
	return <-output
}
